packages/contractkit/src/wallets/ledger-wallet.ts

```
@@ -100,21 +116,47 @@ export class LedgerWallet implements Wallet {
   private async retrieveAccounts() {
     const pairAddresses = await this.retrieveAccountsFromLedger()
     pairAddresses.forEach((pair) =>
-      this.addressDerivationPath.set(normalizeAddressWith0x(pair.address), pair.derivationPath)
+      this.addressDerivationPath.set(normalizeAddressWith0x(pair.address), {
+        derivationPath: pair.derivationPath,
+        validated: pair.validated,
+      })
     )
   }
 
   private async retrieveAccountsFromLedger() {
     const addresses = []
+    const validationRequired = this.validationRequired(false)
     // Each address must be retrieved synchronously, (ledger lock)
     for (const value of this.derivationPathIndexes) {
       const derivationPath = `${this.baseDerivationPath}/${value}`
-      const addressInfo = await this.ledger!.getAddress(derivationPath)
-      addresses.push({ address: addressInfo.address, derivationPath })
+      const addressInfo = await this.ledger!.getAddress(derivationPath, validationRequired)
+      addresses.push({
+        address: addressInfo.address,
+        derivationPath,
+        validated: validationRequired,
+      })
     }
     return addresses
   }
 
+  private validationRequired(initialized: boolean, validated?: boolean): boolean {
+    if (!initialized) {
+      return this.ledgerAddressValidation === AddressValidation.onlyInitialization
+    }
+    switch (this.ledgerAddressValidation) {
+      case AddressValidation.never: {
+        return false
+      }
+      case AddressValidation.everyTransaction: {
+        return true
+      }
+      case AddressValidation.oncePerAddress: {
+        return !validated
+      }
+    }
+    return false
```

We should throw an exception here since it should always be defined

---

Yes, before the onlyInitialization was part of the switch and this was a valid option (but not now, I'll throw an exception)